src/common/tim.rs

```diff
@@ -0,0 +1,71 @@
+use std::{
+    sync::Arc,
+    time::{Duration, Instant}
+};
+
+use crate::rt::{Interval, Sleep, Timer};
+
+/// A user-provided timer to time background tasks.
+pub(crate) type Tim = Option<Arc<dyn Timer + Send + Sync>>;
```

Comment:
I implemented this, but I also want to make sure I understand your thought process fully--it's basically that because someone else has access to the public `Option` and `Arc` types, they might construct their own `Option<Arc<T>> where T: Timer` somewhere and then conveniently take advantage of the fact that they don't have to unwrap it to call the timer methods. And this is basically us leaking an implementation detail of hyper as a piece of externally visible API?

Reply:
Yes, you understand exactly what I was thinking!